### PR TITLE
Compact Block Tweaks

### PR DESCRIPTION
**Pull request by TheBlueMatt**
As I'm avoiding code churn in #8068, any further changes which are not required for correct operation will be dumped here.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -461,9 +461,13 @@ void UpdateBlockAvailability(NodeId nodeid, const uint256 &hash) {
 
 void MaybeSetPeerAsAnnouncingHeaderAndIDs(const CNodeState* nodestate, CNode* pfrom) {
     if (nodestate->fProvidesHeaderAndIDs) {
-        BOOST_FOREACH(const NodeId nodeid, lNodesAnnouncingHeaderAndIDs)
-            if (nodeid == pfrom->GetId())
+        for (std::list<NodeId>::iterator it = lNodesAnnouncingHeaderAndIDs.begin(); it != lNodesAnnouncingHeaderAndIDs.end(); it++) {
+            if (*it == pfrom->GetId()) {
+                lNodesAnnouncingHeaderAndIDs.erase(it);
```
Comment (sipa) [inline src/main.cpp]: This invalidates the 'it' iterator.

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/8235/commits/e0786638d999cdd186cf547cf4f4183c63b2a231

**Comment by sipa** [comment]
tACK e0786638d999cdd186cf547cf4f4183c63b2a231 (tested with the tests from #8418, changing the appropriate `15` and `11` constants in the test to `10` and `6` respectively).

**Comment by sipa** [comment]
Needs rebase after #8418.

**Comment by btcdrak** [comment]
Should this be backported to 0.13?

**Comment by sipa** [comment]
@TheBlueMatt Feel like adapting the tests from #8418 to not fail here?

File: src/main.cpp [hunk 2]
```diff
@@ -4620,7 +4624,7 @@ void static ProcessGetData(CNode* pfrom, const Consensus::Params& consensusParam
                         // they wont have a useful mempool to match against a compact block,
                         // and we dont feel like constructing the object for them, so
                         // instead we respond with the full, non-compact block.
-                        if (mi->second->nHeight >= chainActive.Height() - 10) {
+                        if (CanDirectFetch(Params().GetConsensus()) && mi->second->nHeight >= chainActive.Height() - 5) {
```
Comment (dcousens) [inline src/main.cpp]: Maybe the `5` would make sense as a named constant w/ explanation?

**Comment by laanwj** [comment]
What is the status of this?

**Comment by sipa** [comment]
@laanwj See #8637.

**Comment by TheBlueMatt** [comment]
#8637

**Pull request closed without merging**